### PR TITLE
W2 story follow ups #187175908

### PR DESCRIPTION
**Pull request by tofarr**
* Does not show or link to index page if there is a single invalid W2
* Has additional validation checks on W2s
* Has additional show checks on W2s
* Changes the validation flow on the index page
* Adds specs for all these

### CODE REVIEW

File: app/controllers/state_file/questions/ny_w2_controller.rb [hunk 1]
```diff
@@ -9,14 +9,21 @@ def self.show?(intake)
...
       def index
-        @w2s_with_metadata = @w2s.map do |w2|
-          dfw2 = w2.state_file_intake.direct_file_data.w2s[w2.w2_index]
-          {
-            w2: w2,
-            employer_name: dfw2.EmployerName,
-            wages_amount: dfw2.WagesAmt,
-          }
+        if @w2s.length == 1
+          redirect_to action: :edit, id: @w2s[0].w2_index
```
Comment (rickreyhsig) [inline app/controllers/state_file/questions/ny_w2_controller.rb:13]: If there is only one W2, we don't perform any validation?

File: spec/controllers/state_file/questions/ny_w2_controller_spec.rb [hunk 2]
```diff
@@ -164,6 +164,42 @@
...
+        let(:direct_file_xml) do
+          xml = super()
+          xml.at("LocalIncomeTaxAmt").content = "9000"
+          xml
+        end
+
+        it "returns true" do
+          expect(described_class.show?(intake)).to eq true
+        end
+      end
```
Comment (rickreyhsig) [inline spec/controllers/state_file/questions/ny_w2_controller_spec.rb:202]: Could be good to have a spec where #show returns false also.

File: db/create_analytics_views.sql [hunk 3]
```diff
@@ -297,3 +297,7 @@ CREATE view analytics.state_file_az_intakes AS
 CREATE view analytics.state_file_ny_intakes AS
     SELECT id, updated_at, consented_to_terms_and_conditions, contact_preference, created_at, current_step, eligibility_lived_in_state, eligibility_out_of_state_income, eligibility_part_year_nyc_resident, eligibility_withdrew_529, eligibility_yonkers, hashed_ssn, nyc_maintained_home, nyc_residency, primary_birth_date, primary_esigned_at, primary_state_id_id, raw_direct_file_data, routing_number, sign_in_count, spouse_state_id_id, untaxed_out_of_state_purchases
     FROM public.state_file_ny_intakes;
+
+CREATE view analytics.state_file_w2s AS
+SELECT id, employer_state_id_num, local_income_tax_amt, local_wages_and_tips_amt, locality_nm, state_file_intake_type, state_income_tax_amt, state_wages_amt, w2_index, created_at, updated_at, state_file_intake_id
+FROM public.state_file_w2s;
```
Comment (rickreyhsig) [inline db/create_analytics_views.sql:303]: Is this view for the metabase analytics?

File: app/controllers/state_file/questions/ny_w2_controller.rb [hunk 4]
```diff
@@ -9,14 +9,21 @@ def self.show?(intake)
...
 
       def index
-        @w2s_with_metadata = @w2s.map do |w2|
-          dfw2 = w2.state_file_intake.direct_file_data.w2s[w2.w2_index]
-          {
-            w2: w2,
-            employer_name: dfw2.EmployerName,
-            wages_amount: dfw2.WagesAmt,
-          }
+        if @w2s.length == 1
```
Comment (rickreyhsig) [inline app/controllers/state_file/questions/ny_w2_controller.rb:12]: So this renders the one W2 if there's only one and and index of the W2s if there are more?

**Review by rickreyhsig — COMMENTED** [review]
Overall, the PR makes sense. Just added a few questions for my own sanity prior to approving.

File: app/controllers/state_file/questions/ny_w2_controller.rb (hunk 4 above)
Comment (tofarr) [reply app/controllers/state_file/questions/ny_w2_controller.rb:12]: Correct!

File: app/controllers/state_file/questions/ny_w2_controller.rb (hunk 1 above)
Comment (tofarr) [reply app/controllers/state_file/questions/ny_w2_controller.rb:13]: If there is only 1 W2, we bypass the index page and take users directly into the edit for that W2

File: config/routes.rb [hunk 5]
```diff
@@ -582,7 +582,7 @@ def scoped_navigation_routes(context, navigation)
 
       scope ':us_state', as: 'ny', constraints: { us_state: :ny } do
         scoped_navigation_routes(:questions, Navigation::StateFileNyQuestionNavigation)
-        resources :ny_w2, only: [:index, :edit, :update], module: 'state_file/questions', path: 'questions/ny_w2'
+        resources :ny_w2, only: [:index, :edit, :update, :create], module: 'state_file/questions', path: 'questions/ny_w2'
```
[the remark this answers is not in the document]
Comment (tofarr) [reply config/routes.rb:585]: Only New York has been experiencing this issue - we leave the possiblilty of doing this for other states (Though I would probably rename the controller then from NyW2Controller to simply W2Controller

File: db/create_analytics_views.sql (hunk 3 above)
Comment (tofarr) [reply db/create_analytics_views.sql:303]: Yes - that was one of the requirements of the ticket

File: spec/controllers/state_file/questions/ny_w2_controller_spec.rb (hunk 2 above)
Comment (tofarr) [reply spec/controllers/state_file/questions/ny_w2_controller_spec.rb:202]: Good call - adding it now

Comment (tofarr) [reply spec/controllers/state_file/questions/ny_w2_controller_spec.rb:202]: We have some already on L17, L50 & L112,

**Review by rickreyhsig — APPROVED** [review]

**Pull request merged**